docs/rfcs/2025-06-20-repartition.md

```diff
@@ -0,0 +1,154 @@
+
+Metasrv will trying to update the metadata of partition, or the region route table (related to `PartitionRuleManager`). This step is in the "no ingestion" scope, so no new data will be ingested. Since this won't take much time, the affection to the cluster is minimized. Metasrv will also update the region rule to corresponding regions on Datanodes. 
+
+Every regions and all the ingestion requests to the region server will have a version of region rule, to identify under which rule the request is processed. The version can be something like `hash(region_rule)`. Once the region rule on region server is updated, all ingestion request with old rule will be rejected, and all requests with new rule will be accepted but not visible. They can still be flushed to persisted storage, but their version change (new manifest) will be staged.
+
+Then region 0 (or let metasrv to pick any operational region) will compute the new manifests for all target regions. This step is done by first reading all old manifests, and remapping the files with new partition rule, to get the content of new manifests. Notice this step only handles the manifests before region rule change on region server, and won't touch those staged manifests, as they are already with the new rule.
+
+Those new manifest will be submitted to the corresponding target regions by region 0 via a `RegionEdit` request. If this request falls after a few retries, region 0 will try to rollback this change by directly overwriting the manifest on object storage. and report this failure to metasrv and let the entire repartition procedure to fail. And we can also optionally compute the new manifest for those staged version changes (like another repartition) and submit them to the target regions to make the also visible even if the repartition fails.
+
+In the other hand, a successful `RegionEdit` request also acknowledges those staged version changes and make them visible.
+
+After this step, the repartition is done in the data plane. We can start to process compaction and snapshot again.
+
+## Postprocessing
+
+After the main processing is done, we can do some extra postprocessing to reduce the performance impact of repartition. Including reloading caches in frontend's route table, metasrv's kv cache and datanode's read/write/page cache etc.
+
+We can also schedule an optional compaction to reorganize all the data file under the new partition rule to reduce potential fragmentation or read amplification.
+
+## Procedure
+
+Here describe the repartition procedure step by step:
+
+- <on frontend> Validating repartition request
+- <on frontend> Initialize the repartition procedure
+- Calculate rule diff and repartition plan group
+- Allocate necessary new regions
+- Lock the table key
+- For each repartition subprocedure
+  - Stop compaction and snapshot
```

Comment:
Maybe `manifest checkpoint` is more accurate? Consider replacing snapshot with this in the rfc.
```suggestion
  - Stop compaction and manifest checkpoint
```